### DISCUSSION TITLE
Way to tell if we are at the end of range when the parser is being used as injected

### DISCUSSION THREAD

**Discussion by stsewd** [tree-sitter/tree-sitter#1296] [Q&A] [answered]
Hi, we have a problem on nvim-treesitter where we inject rst on python's docstrings. The parser works correctly when is used without being injected, but it fails to recognize the correct tokens sometimes. RST makes use of the EOF and EOL chars to tell when a dedent or end happens. and I think there it's the problem, since the range of the injection doesn't include those chars.

For example, this will fail to parse correctly (it's being recognized as a title) instead of as a paragraph (this makes me think the parsers is consuming the extra `"""` that are outside the injection range, is this possible? or maybe it's trying to recover from an error?).

```python
class Foo:

    """One two"""
```

But this will be parsed correctly as a paragraph (new line included)

```python
class Foo:

    """One two
    """
```

So, to be able to fix this, is there a way to know from the parser if we are at the end of the range? or to include an extra `EOF` or `EOL` at the end of the range so parsers that depend on them can work correctly?

I have tried using

 https://github.com/tree-sitter/tree-sitter/blob/d63bbbaadd597e3bb78d55f94d7e0ae0f8c39b04/lib/src/lexer.c#L39-L45

but the problem is still present.

To confirm this, I have added `\` as an EOF char, and by doing so this works

```python
class Foo:

    """One two\\"""
```

**Comment by stsewd** [comment, marked as answer]
Looks like injections do add an `EOF` mark, but in my parser I was always assuming that it would be accompanied by an `\n`. Fixed this in my parser with https://github.com/stsewd/tree-sitter-rst/commit/41a15b8d530cc7262ff56a3a3ecdfe3f4ddec120

**Reply by ahlinc** [reply]
I'd suggest to use the `lexer->is_eof()` call instead of the `lexer->lookahead == 0` comparison, cause both do the same, but I suppose the second may be a subject to change in the future.
